extensions/machine-learning-services/src/common/queryRunner.ts

```diff
@@ -57,8 +71,36 @@ export class QueryRunner {
 	 * @param connection SQL Connection
 	 */
 	public async getPythonPackages(connection: azdata.connection.ConnectionProfile): Promise<nbExtensionApis.IPackageDetails[]> {
+		return this.getPackages(connection, listPythonPackagesQuery);
+	}
+
+	/**
+	 * Returns python packages installed in SQL server instance
+	 * @param connection SQL Connection
+	 */
+	public async getRPackages(connection: azdata.connection.ConnectionProfile): Promise<nbExtensionApis.IPackageDetails[]> {
+		return this.getPackages(connection, listRPackagesQuery);
+	}
+
+	/**
+ * Returns python packages installed in SQL server instance
+ * @param connection SQL Connection
+ */
+	public async getRAvailablePackages(connection: azdata.connection.ConnectionProfile): Promise<nbExtensionApis.IPackageDetails[]> {
+		return this.getPackages(connection, listRAvailablePackagesQuery);
+	}
+
+	private async getPackages(connection: azdata.connection.ConnectionProfile, script: string): Promise<nbExtensionApis.IPackageDetails[]> {
 		let packages: nbExtensionApis.IPackageDetails[] = [];
-		let result = await this.runQuery(connection, listPythonPackagesQuery);
+		let result: azdata.SimpleExecuteResult | undefined = undefined;
+
+		for (let index = 0; index < 3; index++) {
+			result = await this.runQuery(connection, script);
```

Comment:
Why 3 times? Do you expect this to fail often or something in a way that the loop will continue to execute and will return a result in the subsequent executions?

Reply:
it needs retry. not sure how many is enough right now

---
In reply to: [366614028](https://github.com/microsoft/azuredatastudio/pull/8870#discussion_r366614028) [](ancestors = 366614028)

Reply:
Why does it need a retry? And if it does shouldn't you be doing a back-off strategy? It'd seem unlikely to me that if it failed the first time that executing it again immediately 2 more times is going to solve it. Shouldn't you at least be waiting a second or two between attempts?

Reply:
I see sometimes the query doesn't return anything. if I retry it works correctly, I'm not sure why I have to investigate 

---
In reply to: [366654864](https://github.com/microsoft/azuredatastudio/pull/8870#discussion_r366654864) [](ancestors = 366654864)